Add EIP: Increase max attestation inclusion slot (#7045)
ethereum/EIPs: EIPS/eip-7045.md

---
eip: 7045
title: Increase max attestation inclusion slot
description: Increases max attestaton inclusion slot to the last slot in `N+1` where `N` is the epoch containing the attestations slot.
author: Danny Ryan (@djrtwo)
discussions-to: https://ethereum-magicians.org/t/eip-7045-increase-attestation-slot-inclusion-range/14342
status: Draft
type: Standards Track
category: Core
created: 2023-05-18
---

## Abstract

> Increases max attestaton inclusion slot from `attestation.slot + SLOTS_PER_EPOCH` the last slot in `N+1` where `N` is the epoch containing the attestations slot.

Reviewer: ```suggestion
Increases max attestation inclusion slot from `attestation.slot + SLOTS_PER_EPOCH` to the last slot of epoch `N+1` where `N` is the epoch containing the attestation slot.
```

This increase is critical to the current LMD-GHOST security analysis as well as the confirmation rule.

## Motivation

Attestations can currently be included after some minimum delay (`1` slot on mainnet) up until `SLOTS_PER_EPOCH` slots after the slot the attestation was created in. This rolling window of one epoch was decided upon during Phase 0 because the equal inclusion window for any attestation was assessed as "fair". The alternative considered path was to allow inclusion during the current and next epoch which means attestations created during the start of an epoch have more potential slots of inclusion than those at the end of the epoch.

> Since this decision was initially made, it has become apparent that the alternative design is critical for current LMD-GHOST security proofs as well as a new confirmation rule (which will allow for block confirmations in approximately 3-4 slots in normal mainnet conditions).

Reviewer: ```suggestion
Since this decision, it has become apparent that the alternative design is critical for current LMD-GHOST security proofs as well as a new confirmation rule (which will allow for block confirmations in approximately 3-4 slots in normal mainnet conditions).
```

This specification, thus increases the max inclusion slot for attestations in accordance with the learned security proof and confirmation rule needs.

## Specification

### Constants

| Name | Value | Comment |
| - | - | - |
|`FORK_TIMESTAMP` | *TBD* | Mainnet |

### Execution layer

This requires no changes to the Execution Layer.

### Consensus layer

The specification makes two minor changes to the state transition function:

* Modify `process_attestation` to not have an upper bound on the slot check and instead just rely on the minimum slot as well as the target epoch being in either current or previous epoch.
* Modify `get_attestation_participation_flag_indices` to set the `TIMELY_TARGET_FLAG` without consideration of `inclusion_delay` to ensure that the extended inclusion attestations have a non-zero reward.

Additionally, the specification modifies the attestation and aggregate attestation gossip conditions to allow for gossip during this extended range.

## Rationale

### Extended max inclusion slot

> As discussed in the Motivation, extending this max inclusion slot the end of the next epoch is critical for LMD-GHOST security proofs and confirmation rule.

Reviewer: ```suggestion
As discussed in the Motivation, extending this max inclusion slot to the end of the next epoch is critical for LMD-GHOST security proofs and confirmation rule.
```

### Removal of `inclusion_delay` consideration for target reward

Previously, `get_attestation_participation_flag_indices` would only set the `TIMELY_TARGET_FLAG` (and thus reward for an attestation with correct target vote) if the attestation was included within a `SLOTS_PER_EPOCH` window.

> The `inclusion_delay` consideration for this flag is removed to ensure that whatever the valid inclusion window is for an attestation that an attestation can receive a baseline non-zero reward for correct target. This ensures that clients will still attempt to pack such attestations into blocks which is important for the security analysis.

Reviewer: ```suggestion
The `inclusion_delay` consideration for this flag is removed to ensure that whatever the valid inclusion window is for an attestation, it can receive a baseline non-zero reward for correct target. This ensures that clients will still attempt to pack such attestations into blocks which is important for the security analysis.
```

## Backwards Compatibility

This EIP introduces backwards incompatible changes to the block validation rule set on the consensus layer and must be accompanied by a hard fork.

## Security Considerations

> This improves LMD-GHOSt security as well as enabling a fast confirmation rule.

Reviewer: ```suggestion
This improves LMD-GHOST security as well as enables a fast confirmation rule.
```
